### ISSUE TITLE
All Nim versions since 1.2 up to devel: `--gc:orc` memory leak for closure captured `ref`'s...

### ISSUE THREAD

**Issue by GordonBGood** [nim-lang/Nim#18420]
Use of a cyclic reference structure where --gc:orc is truly needed leaks memory for some uses of combinations of `--gc:orc` options, especially the new undocumented `-d:nimThinout`.

The problem seems to be related to breaking of cycles for captured closure references; thus, it will apply in many practical uses.  It may be caused by nested closures, but that should be a possible use case.

The fact that `--gc:orc` doesn't work reliably for versions 1.2.x is likely not so much a concern, as this feature was only experimental at that stage, but for versions 1.4.x where the feature was officially released, it should work consistently where it does not.

The other variable is the use of the new undocumented (experimental?) `-d:nimThinout` flag, which seems to be a method to shorten and automate the cycle breaking scans so that no manual calling of `GC_FullCollect()` needs be made (in fact, this proc isn't even available when the flag is enabled).

### Example

The problem was identified in using Co-Inductive Streams (CIS's) and/or Lazy List reference objects, which must necessarily contain a "thunk" closure that is called to defer (lazy) calculate the tail of the stream/list, with the difference between the two that the stream is not memoized where lazy lists are.

Thus, the same problem is present in Lazy List's used in [the current Hamming Numbers Orc test](https://github.com/nim-lang/Nim/blob/8ccde68f132be4dba330eb6ec50f4679e564efac/tests/arc/thamming_thinout.nim) if it is run, but the following example code using CIS's reduces and simplifies the requirement to perhaps make it easier to identify the problem:
The following mini program is cyclic by design so one compiles with `--gc:orc` and it seems that (undocumented) `-d:nimThinout` is also needed, else the `getOccupiedMem()` measured memory leak is atrocious (why does it not use this flag by default?):
```nim
import sugar

type CIS = ref object # a custom CIS of int's...
       head: int
       tailf: proc(): CIS {.closure.}

proc nxt(cis: CIS): CIS = # variable recursion (shared fix point)
  var r: CIS = nil # recursion target
  r = CIS(head: cis.head + 1, tailf: () => r.nxt)
  r

iterator naturalsFor(count: int): int =
  var nats: CIS = nil # so we have a call target!
  nats = CIS(head: 0, tailf: () => nats.nxt)
  for _ in 1 .. count:
    yield nats.head
#    nats = nats.tailf() # desired code - the bigger error using this
# [ # remove the space to comment out the following...
    var f = nats.tailf
#    nats.tailf = nil # with this, the error is only one allocation
    nats = f()
#  nats.tailf = nil # and with this, we remove the last allocation manually
# ]#

proc main =
  stdout.write "The first 10 natural numbers are:  "
  for n in naturalsFor 10: stdout.write n, " "
  echo ""

let mem = getOccupiedMem()
main()
# GC_FullCollect() # this makes it work, but doesn't work with -d:nimThinout
echo "Memory balance:  ", getOccupiedMem() - mem
```

As shown above, manually calling a full memory collection does actually work with the latest devel version using just `--gc:orc` but it leaves the programmer to decide where memory needs to be recovered and thus, isn't the same programmer interface as `--gc:arc` where the recovery is automatic when the references go out of scope.  The idea of the `-d:nimThinout` to make the memory collection automatic is brilliant if it only worked in all cases.

### Current Output
```
The first 10 natural numbers are:  0 1 2 3 4 5 6 7 8 9 
Memory balance:  960
```

### Expected Output
```
The first 10 natural numbers are:  0 1 2 3 4 5 6 7 8 9 
Memory balance:  0
```

### Possible Solution

Note that if run with `--gc:arc` (even though it is known that the references are cyclic) the actual memory leak is only two thirds of the use as specified above; however, this is likely just due to a couple of words of storage overhead for the "Orc/thinout" use.

A clue might be that if the internal closure "thunk" is `nil`'ed out manually rather than letting the run time default handle it, it removes the memory leak; however, that is the whole point of "Orc" to make manual breaking of cycles unnecessary!

It's possible that this leak is related to the use being inside a closure iterator: ie. nested closures...

### Additional Information
It looks like this has never entirely worked properly, or if it did then back ported later changes have caused the current problem in all versions since `--gc:orc` has been an option.

This is a showstopper issue for the future general use of --gc:orc covering all cyclic data cases!!!

```
$ nim -v
Nim Compiler Version 1.2.0 **and up to latest devel as of 2 July 2021** [Linux: amd64]; latest tried:
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-07-03
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: d1d2498c7b933b6364af450df5088aaa89b4cf4b
active boot switches: -d:release
```

**Comment by GordonBGood** [comment]
In fact, as I speculated in a comment on issue #18421, these two issues are likely related.  If I change the definition of the `CIS` type to the following:
```nim
type
  CISObj = object # a custom CIS of int's...
    head: int
    tailf: proc(): CIS {.closure.}
  CIS = ref CISObj

proc `=destroy`(co: var CISObj) =
  echo "destroying CIS"
  if co.tailf != nil: co.tailf.`=destroy`
  wasmoved(co.tailf)
  wasmoved(co)
```
then the output is as follows no matter if `-d:nimThinout` is used or not (except that the memory balance is zero as desired when collection is "forced" with `GC_FullCollect()` without `-d:nimThinout`):
```
The first 10 natural numbers are:  0 destroying CIS
1 2 3 4 5 6 7 8 9 
Memory balance:  XXX
```
Note that the destructor is called only once for the the destruction of the head of the entire CIS chain and then never called again; this indicates to me that once the first cycle is detected, something changes in the destructor logic so further internal destructors are not called, both for the forced "GC_FullCollect90` and for the new "Thinout" algorithm.

I think that if this bug were fixed, there might be no need to call `GC_FullCollect` when `-d:nimThinout` is not used as (hopefully) the cycles would be broken and the destructors run as the chain of CIS's is consumed.

When neither `-d:nimThinout` nor `GC_FullCollect()` is used, the very high memory balance indicated of 21440 is likely just an artifact of `gotoccupiedMem()` finding redundant memory chains due to tracing the cycles.

**Comment by GordonBGood** [comment]
I am informed in the related issue #18421 that `-d:nimThinout` is a "failed experiment" and shouldn't be used. although (at least in the use as for here) it does actually "thin out" the amount of occupied memory measured by `getOccupiedMem()` by eliminating the redundant counts caused by the cycles.  I wonder if the experiment may have failed due to the problem we have identified here and in the related issue in that the destruction of graphs that have been determined to contain cycles are deferred to some determined collection point, whether manually by calling `GC_FullCollect()` as here or at some determined memory level which is missing calling some destructors (or perhaps only custom destructors).  Anyway, that's something that only @araq knows, so I'll leave it.

So then the question is whether there is an issue other than as also found in that related issue:  For me, it seems that there is in that the measured occupied memory is completely inaccurate so that collection cycles will be triggered much more often than they need to be, but it probably isn't a Showstopper...

**Comment by Araq** [comment]
`-d:nimThinout` is manual cycle break-up, it's not "brilliant", it is a failed experiment -- it couldn't handle async's cycles. (Async is the only real reason why I care about a solution for cycles. ;-) )

**Comment by GordonBGood** [comment]
@araq:

>  `-d:nimThinout` is manual cycle break-up ... it couldn't handle async's cycles.

Now I understand.

> Async is the only real reason why I care about a solution for cycles. ;-) 

Oh, I don't know, I didn't think I cared either, but it turns out that the Hamming Numbers functional algorithm is cyclic (when we were working with it before, I didn't think it was) so there are likely other similar things that need Orc's cycle breaker.

Since the Hamming Number Test turns out to be cyclic, it obviously can't work with Arc (memory leaks), but I was surprised to find that `getOccupiedMem()` shows a memory leak using the same test method with the default GC.  I haven't investigated that much as we are trying to move toward using Orc by default, but I suppose we need to support legacy code and LTE versions?  Or maybe I'm missing something and we don't expect `getOccupiedMem()` to balance before and after even with a forced `GC_FullCollect()?

**Comment by Araq** [comment]
> Or maybe I'm missing something and we don't expect getOccupiedMem() to balance before and after even with a forced `GC_FullCollect()?

We don't expect this to be true for the old GCs indeed. But we do expect it for --gc:orc.

**Comment by GordonBGood** [comment]
@araq:
> We don't expect (`getOccupiedMem()` memory balance) to be true for the old GCs indeed. But we do expect it for --gc:orc.

Good to know, thanks.

> -d:nimThinout is manual cycle break-up, it's not "brilliant", it is a failed experiment -- it couldn't handle async's cycles ...

It seems you were pretty excited about it when you wrote it, going by the comment header ;-).  I liked it because it seemed to give an accurate measure of the occupied memory when enabled, but I didn't understand that it is a "manual cycle break-up" so that calls to `thinout()` needed to be made at the appropriate times to do the actual destruction of the references it handled.  I suppose it's a "failed experiment" because you couldn't find a way to have the compiler inject these calls automatically at the times that destructors would normally be called and still have the code work for all cases.

If calling it at the wrong time could cause premature destruction, then it would be worse than a memory leak as it would cause the code, in this case "async", not to work at all.

**Comment by Araq** [comment]
So ... Duplicate of https://github.com/nim-lang/Nim/issues/18421 ?

**Comment by GordonBGood** [comment]
@araq:

> So ... Duplicate of #18421 ?

**Yes, it has evolved** that @Clyybber's issue #18421 is a duplicate of this issue, but **I like the title of his better** as it better describes the real problem ;-).

Given that I misunderstood the usefulness of `-d:nimThinout` and also that `GC_FullCollect()` must be called before checking the memory balance, **please close this issue after answering the following two questions (if you would)**:

1. Would it be possible to **modify `getOccupiedMem()`** when using it with `--gc:orc` **so it gives a more reasonable answer** such as 640 for the above program if `--gc:arc` is used (which can't be collected because they are cyclic references), 960 when `--gc:orc` and (defunct) `-d:nimThinout` was used (which could only be collected by manually calling `thinout` on the base reference), instead of 21440 before `GC_FullCollect()` is called when `--gc:orc` is used by itself?  I understand that this number is hugely inflated by redundant tracing due to the multiple cycles and it drops to zero after collection is forced, of course.  However, for the programmer who tries to actually use the number to try to trace what is using memory, this huge number is completely useless.  I can't see why it doesn't give one of the other two answers (which likely depend on if there are some extra hidden fields when using Orc `ref`'s).

2. Would it be possible to **automatically collect** as in a `GC_FullCollect(X)` (or equivalent to `thinout(X)`) on **the whole group of cyclic references within "X" when "X" goes out of scope, when "X" does not contain any cyclic references that are not within its bounds?**

**Comment by Araq** [comment]
1. `getOccupiedMem` doesn't know about the GC mode and gives precise numbers. The number includes the size of the array of "cycle candidates" and that's probably why you see such a high number.

2. Looks hard to do and worse, more features don't make the system as a whole easier to understand.

**Comment by GordonBGood** [comment]
@Araq:

Thank you for your answers; understand about the "cycle candidates" and appreciate your candid thoughts on the second question.  Closing now.
